lib/router/src/lib.rs

```diff
@@ -208,43 +260,113 @@ where
     fn not_recognized() -> Self {
         Self::error(error::NotRecognized.into())
     }
-
-    fn no_capacity(capacity: usize) -> Self {
-        Self::error(error::NoCapacity(capacity).into())
-    }
 }
 
-impl<Req, Svc> Future for ResponseFuture<Req, Svc>
+impl<Req, Rec, Mk> Future for ResponseFuture<Req, Rec, Mk>
 where
-    Svc: svc::Service<Req>,
-    Svc::Error: Into<error::Error>,
+    Rec: Recognize<Req>,
+    Mk: Make<Rec::Target>,
+    Mk::Value: svc::Service<Req> + Clone,
+    <Mk::Value as svc::Service<Req>>::Error: Into<error::Error>,
 {
-    type Item = <LoadShed<Svc> as svc::Service<Req>>::Response;
+    type Item = <LoadShed<Mk::Value> as svc::Service<Req>>::Response;
     type Error = error::Error;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         use svc::Service;
 
         loop {
             self.state = match self.state {
-                State::Init(ref mut req, ref mut svc) => {
-                    match svc.poll_ready().map_err(error::Error::from)? {
-                        Async::NotReady => {
-                            unreachable!("load shedding services must always be ready");
+                State::Init(ref mut state) => {
+                    let request = state.request.take().expect("polled after ready");
+                    let target = state.target.take().expect("polled after ready");
+                    let make = state.make.take().expect("polled after ready");
```

Comment:
This is a bug ;)  If the lock is not ready, you'll have taken these values and they won't be available on the next invocation!  This should happen after you've acquired the cache.

Reply:
Ah good call, thanks! I have a few changes to roll up so I’ll make sure to fix this.